### DISCUSSION TITLE
General Intellistep Performance Observations

### DISCUSSION THREAD

**Discussion by CAP1Sup** [CAP1Sup/Intellistep#7] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.
 
 **Note that this is my first time using discussions, let's see how it goes.**

**Comment by lav8org** [comment]
Introduction, ah, +\- yet another weirdo hacker type; got my start in earnest diy fungineering with the 2nd gen makerbot (ToM), also have a background in ML and graphics. My goto "something about yourself": rode a bicycle across the US once upon a time (highly recommended).

My Intellistep relevant project is a souped up plotter/drag engraving robot, trying to place .01mm lines with roughly that positioning accuracy/precision at a rate sufficient to cut millions of lines in less than days. I've just run my first few tests with the stock S42B fw and the results were... hilaribad, the worst I've seen. One of my test patterns is a Hilbert curve (maximum torturous tortuousness) and with TMC 5160s running open loop I'd get irregular gap widths that appear to be a very small dead reckoning drift... well the gaps _are_ gone now, but so is any semblance of the actual curve 
![IMG_20210403_130818](https://user-images.githubusercontent.com/46395752/113489225-3ae8eb80-9480-11eb-93a6-4970f93aa028.jpg)

idk where intellistep falls on the tune up/rewrite spectrum, but if it's anywhere near the former I think it's worth noting that the steps/mm is suspiciously off. I calibrate my axes with a dial indicator and the correct values differed from those for the 5160s by a significant amount (factor of 10/9)
I think the rest of what's on my mind fits better in the other topic...

**Reply by CAP1Sup** [reply]
Intellistep uses redone stepping, that should fix the way that BTT had their steps/mm. Steps/mm should match exactly with any other standard driver once everything is working properly. I want to make the conversion process as smooth as possible.

**Comment by xerootg** [comment]
Welcome to the strange hacker collective! So on your calibration, keep in
mind that the s42b controllers interpolate one step across a binary
multiple regardless of natural steps of the stepper. For example, in 32 bit
mode, you have 8192 steps per rotation of the stepper. The 5160 on the
other hand, depending on what chopping mode you use, will average steps out
in the name of noise. It's kind of surprising to see your side-by-side
there, but I suspect that you could benefit from some pid tuning for your
application. This project does not have functioning pid tuning right now,
but it is definitely in the works to implement positional pid instead if
speed control pid. If your running the stock s42b firmware, you can use the
serial interface to play with pid settings.

**Reply by lav8org** [reply]
Ahh, at last, my people! 
Yeah for calibration I sidestep the perils of tiny sample sizes by measuring the largest manageable throw (good that I do, I noticed for the first time a .05 backlash on one axis, probably tooth deflection due to a toothless idler, which could've thrown off the calibration were it against a small throw). For the DI 10mm was convenient, hence the 10/9 factor. Actually 2 of the 3 axes were off by a factor of 10/8 but unlike the former I'm not totally sure I had them dialed in, and if I did it was against 16x microstepping vs 32 now... enough uncertainty to muddy the waters and I'd hate to waste time by throwing out red herrings.

If it's not the only user facing parameters I suppose it's just the firmware that's garbage, eh? Can't say I'd be surprised, since I've noticed some other questionable things like the motors never actually disabling; it seems the current is just reduced instead, since the axles turn pretty easily but they're still whining, 
Anyway, I think I'll start with trying open loop mode to eliminate other possibilities, I must admit I have more practice at flipping a switch than tuning PID params,

**Reply by lav8org** [reply]
open loop ran fine... currently searching for a way to get serial through stlink on linux, not as straightforward as it appeared.

**Reply by CAP1Sup** [reply]
> open loop ran fine... currently searching for a way to get serial through stlink on linux, not as straightforward as it appeared.

Honestly, I don't think it's possible. I use a cheap FTDI-232 board, that seems to be the best route.

**Reply by lav8org** [reply]
> Honestly, I don't think it's possible. I use a cheap FTDI-232 board, that seems to be the best route.

That certainly would explain why I was having so much trouble! 
So I saw "*-232" and that it's a usb/serial adapter and jumped to a dubious conclusion, hacked together a dsub breakout pigtail for an old adapter I have, and couldn't get more than a break upon connect. _Then_ decided to take the two minutes to search if this wasn't foolhardy. Fingers crossed my confident ignorance didn't lead to legacy rs-232 voltages frying the ttl interface (the magic smoke is still contained, so at least there's hope). Seems I'll have to look for other things to break until the FTDI-232 arrives.

**Reply by CAP1Sup** [reply]
> > Honestly, I don't think it's possible. I use a cheap FTDI-232 board, that seems to be the best route.
> 
> 
> 
> That certainly would explain why I was having so much trouble! 
> 
> So I saw "*-232" and that it's a usb/serial adapter and jumped to a dubious conclusion, hacked together a dsub breakout pigtail for an old adapter I have, and couldn't get more than a break upon connect. _Then_ decided to take the two minutes to search if this wasn't foolhardy. Fingers crossed my confident ignorance didn't lead to legacy rs-232 voltages frying the ttl interface (the magic smoke is still contained, so at least there's hope). Seems I'll have to look for other things to break until the FTDI-232 arrives.

If you can't wait, you could probably use a spare Arduino. They work with similar serial converters. You could connect pins 0 and 1 to the RX/TX pins of the stepper. That should work for you till you get your FTDI board

**Reply by TazerReloaded** [reply]
An Arduino can be used to do serial passthrough. Pin 0 and 1 are connected to the USB serial chip. You can just attach jumpers from there to the stepper driver. The Arduino should ideally have no sketch installed so it does not respond to any commands. On bigger Arduinos like the Mega, you have multiple serial interfaces, a little pass-through sketch could connect them together. You could also try software serial, but that tends to be quite unreliable.

**Comment by GhostlyCrowd** [comment]


**Reply by CAP1Sup** [reply]
Yup, it's stuck trying to initialize the SPI interface with the encoder. I'm still working through the issues. I'm currently getting a CRC error. Hopefully I can fix it later today, if not I don't have a timeline for when I'll be able to fix it. I'm hoping as soon as possible (probably a week or two) but this is literally the last major issue that I'm struggling with. If anyone wants to take a crack at it, be my guest. I fixed the current issue of the speed being too high but otherwise I'm still trying to fix the CRC issue. I'll push later tonight with the speed fix. All the encoder interface code is run in the encoderInit() function in encoder.cpp

**Reply by GhostlyCrowd** [reply]
No worries, just figured id post it just in case.

**Reply by CAP1Sup** [reply]
Yup, this is fixed. Encoder values are present as well

**Comment by GhostlyCrowd** [comment]
Got caught up today. Stepper is enabled now, but it has a high-pitched whine, no stepper movement, and the stepper gets warm pretty rapidly. I'm assuming your aware since there is more to be done, but I figured it post just in case.

**Reply by CAP1Sup** [reply]
I wasn't aware of it, although I knew that the stepping needed testing. I have a test setup that will allow me to continue to program the board while the power supply is connected to it. I'm going to try to make some time tonight to start debugging the issues.

**Comment by CAP1Sup** [comment]
0.0.23 Observations

**Reply by CAP1Sup** [reply]
@TazerReloaded @GhostlyCrowd @lav8org @jlmxyz would you guys mind testing out the code on your boards? As of right now, it should be fully set up for basic functions (moving, holding position, etc). It also sounds really bad, something that I'll be looking into. However, everything seems to be working properly from my perspective. I'm going to do my own testing but I figured that it would help to have some extra eyes. As it is configured, it should be a drop in replacement for a 16th microstepping driver, with the BTT max pulse limiting as before. The CAN is completely untested, but the serial should work as well. Dynamic current might work, but unlikely. I'm getting close to bringing all of the functionality I planned, and my schedule is open, so keep a heads up as I'll be updating the code much more frequently.

**Reply by GhostlyCrowd** [reply]
I should have some free time tonight after work.

**Comment by CAP1Sup** [comment]
I was forced to put my test motor onto my printer with version 0.0.42 due to lost parts on my R/C car. The parts were small, plastic parts that the car cannot run without, but would cost a ton to ship. I put the Intellistep board on the X, original BTT firmware on the Y. So far, the Intellistep board as been performing flawlessly. No visible layer shifts (I'm using PID correction). I did see that there is additional lag between the motor getting up to speed, but I think that's only the stock firmware. The motor is also nice and quiet and hasn't heated up past around 45-50 deg. Honestly, I'm really happy with the performance so far. Good inspiration to continue working. Thanks to everyone who's helped out, especially @IhorNehrutsa , @TazerReloaded , @xerootg , and @GhostlyCrowd for code work as well as testing

**Comment by jlmxyz** [comment]
hi,
sorry for late answer
is there any prebuilt I can test with?
thanks and regards

Le ven. 9 juil. 2021 à 21:41, Christian Piper ***@***.***> a
écrit :

**Reply by CAP1Sup** [reply]
Unfortunately, not yet. There will be in the releases, but for now, you'll have to download VSCode and PlatformIO and compile it there
